File: pkg/spec/v1alpha1/environment.go
package v1alpha1

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// New creates a new Environment object with internal values already set
func New() *Environment {
	c := Environment{}

	// constants
	c.APIVersion = "tanka.dev/v1alpha1"
	c.Kind = "Environment"

	// default namespace
	c.Spec.Namespace = "default"

	c.Metadata.Labels = make(map[string]string)

	return &c
}

// Environment represents a set of resources in relation to its Kubernetes cluster
type Environment struct {
	APIVersion string      `json:"apiVersion"`
	Kind       string      `json:"kind"`
	Metadata   Metadata    `json:"metadata"`
	Spec       Spec        `json:"spec"`
	Data       interface{} `json:"data,omitempty"`
}

// Metadata is meant for humans and not parsed
type Metadata struct {
	Name      string            `json:"name,omitempty"`
	Namespace string            `json:"namespace,omitempty"`
	Labels    map[string]string `json:"labels,omitempty"`
}

// Has and Get make Metadata a simple wrapper for labels.Labels to use our map in their querier
func (m Metadata) Has(label string) (exists bool) {
	_, exists = m.Labels[label]
	return exists
}

// Get implements Get for labels.Labels interface
func (m Metadata) Get(label string) (value string) {
	return m.Labels[label]
}

func (m Metadata) NameLabel() string {
	partsHash := sha256.Sum256([]byte(fmt.Sprintf("%s:%s", m.Name, m.Namespace)))
	chars := []rune(hex.EncodeToString(partsHash[:]))
	return string(chars[:48])
}

// Spec defines Kubernetes properties
type Spec struct {
	APIServer        string           `json:"apiServer"`
	ContextNames     []string         `json:"contextNames"`
	Namespace        string           `json:"namespace"`
	DiffStrategy     string           `json:"diffStrategy,omitempty"`
	InjectLabels     bool             `json:"injectLabels,omitempty"`
	ResourceDefaults ResourceDefaults `json:"resourceDefaults"`
	ExpectVersions   ExpectVersions   `json:"expectVersions"`
}

// ExpectVersions holds semantic version constraints
// TODO: extend this to handle more than Tanka
type ExpectVersions struct {
	Tanka string `json:"tanka,omitempty"`
}

// ResourceDefaults will be inserted in any manifests that tanka processes.
type ResourceDefaults struct {
	Annotations map[string]string `json:"annotations,omitempty"`
	Labels      map[string]string `json:"labels,omitempty"`
}

File: rust-export/src/environment.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Environment {
    pub api_version: String,
    pub kind: String,
    pub metadata: Metadata,
    pub spec: Spec,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

impl Environment {
    pub fn new() -> Self {
        Environment {
            api_version: "tanka.dev/v1alpha1".to_string(),
            kind: "Environment".to_string(),
            metadata: Metadata::default(),
            spec: Spec::default(),
            data: None,
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Metadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub namespace: Option<String>,
    #[serde(default)]
    pub labels: HashMap<String, String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Spec {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_server: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_names: Option<Vec<String>>,
    #[serde(default = "default_namespace")]
    pub namespace: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub diff_strategy: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub apply_strategy: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub inject_labels: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tanka_env_label_from_fields: Option<Vec<String>>,
    #[serde(default)]
    pub resource_defaults: ResourceDefaults,
    #[serde(default)]
    pub expect_versions: ExpectVersions,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub export_jsonnet_implementation: Option<String>,
}

fn default_namespace() -> String {
    "default".to_string()
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ResourceDefaults {
    #[serde(default)]
    pub annotations: HashMap<String, String>,
    #[serde(default)]
    pub labels: HashMap<String, String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ExpectVersions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tanka: Option<String>,
}

/// Represents a Tanka environment with its file path
#[derive(Debug, Clone)]
pub struct LoadedEnvironment {
    pub path: PathBuf,
    pub environment: Environment,
}
